### PR TITLE
Time-to-first-frame metrics benchmarking and timeline chart

### PR DESCRIPTION
**Pull request by niyatim23**
*Issue #, if available:*

*What was changed?*
Introduced a new feature which allows receiving master-side metrics in the JS viewer

*Why was it changed?*
It would be easier to understand the time-to-first-frame and which process on both sides is time-consuming and how much time does each API call take

*How was it changed?*
- The master-side (KVS WebRTC C SDK) metrics are sent to the viewer in multiple JSON messages via the datachannel opened by the viewer
- The datachannel first sends a message of the following format with t1 attached, to which the master responds back with the same message attaching t2. In response to this, the viewer sends the same message back with t3 and so on until t5. The viewer is responsible for attaching t1, t3, t5. The master is responsible for t2 and t4. (Master e2e time: t4 - t2, Viewer e2e time: t3 - t1):
```
let dataChannelLatencyCalcMessage = {
    content: 'Opened data channel by viewer',
    t1: '',
    t2: '',
    t3: '',
    t4: '',
    t5: ''
}
```

*What testing was done for the changes?*
Tested the changes locally by running the master first and the viewer first to see that the plot had all the relevant values

Viewer is started first
Screenshot 2023-11-14 at 9 10 02 AM

Master is started first
Screenshot 2023-11-14 at 9 10 47 AM

Tooltip
Untitled

Datachannel messages from master
Screenshot 2023-11-14 at 9 11 16 AM

Checkbox
Screenshot 2023-12-19 at 9 00 54 PM

Timer
Screenshot 2023-12-19 at 8 56 01 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/viewer.js [hunk 1]
```diff
@@ -646,3 +1022,76 @@ function calcStats(stats, clientId) {
...
+    return time > minTime ? new Date(time - diffInMillis) : new Date(minTime - diffInMillis);
+}
+
+function drawChart() {
+    let viewerOrder = ['signaling', 'describeChannel', 'describeMediaStorageConfiguration', 'channelEndpoint', 'iceServerConfig', 'connectAsViewer', 'setupMediaPlayer', 'waitTime',
+                    'offAnswerTime', 'iceGathering', 'peerConnection', 'dataChannel', 'ttffAfterPc', 'video'];
+    let masterOrder = ['signaling', 'describeChannel', 'channelEndpoint', 'iceServerConfig', 'getToken', 'createChannel', 'connectAsMaster', 'waitTime', 
+                    'offAnswerTime', 'iceGathering', 'peerConnection', 'dataChannel', 'ttffAfterPc'];
+    let container = document.getElementById('timeline-chart');
+    let rowHeight = 45;
```
Comment (sirknightj) [inline examples/viewer.js:1043]: Let's use `const` for the constants

File: examples/viewer.js [hunk 2]
```diff
@@ -646,3 +1022,76 @@ function calcStats(stats, clientId) {
...
+    let colors = [];
+
+    dataTable.addColumn({ type: 'string', id: 'Term' });
+    dataTable.addColumn({ type: 'string', id: 'Bar label' });
+    dataTable.addColumn({ type: 'string', role: 'tooltip' });
+    dataTable.addColumn({ type: 'date', id: 'Start' });
+    dataTable.addColumn({ type: 'date', id: 'End' });
+
+    masterOrder.forEach((key) => {
+        if (metrics.master[key] != null || metrics.master[key] != undefined) {
```
Comment (sirknightj) [inline examples/viewer.js:1058]: Let's use `!==` over `!=`. Also, should this be a `&&`?

File: examples/viewer.js [hunk 3]
```diff
@@ -223,18 +473,122 @@ async function startViewer(localView, remoteView, formValues, onStatsReport, onR
...
+        
+                            metrics.master.connectAsMaster.startTime = dataChannelMessage.connectStartTime;
+                            metrics.master.connectAsMaster.endTime = dataChannelMessage.connectEndTime;
+                            
+                        } else if (dataChannelMessage.hasOwnProperty('candidateGatheringStartTime')) {
+                            metrics.master.iceGathering.startTime = dataChannelMessage.candidateGatheringStartTime;
+                            metrics.master.iceGathering.endTime = dataChannelMessage.candidateGatheringEndTime;
+                        } 
+                    } catch (e) {
+                        console.log("Sending a non-json message");
```
Comment (sirknightj) [inline examples/viewer.js:583]: Received a non-json message? (instead of sending)

File: examples/viewer.js [hunk 4]
```diff
@@ -223,18 +473,122 @@ async function startViewer(localView, remoteView, formValues, onStatsReport, onR
...
-            dataChannelObj.onmessage = onRemoteDataMessage;
+            let onRemoteDataMessageViewer = (message) => {
+                
+                remoteMessage.append(`${message.data}\n\n`);
+                if (formValues.enableDQPmetrics) {
+                    try {
+                        let dataChannelMessage = JSON.parse(message.data);
+                        if (dataChannelMessage.hasOwnProperty('t1')) {
+                            test = dataChannelMessage;
+                            if (dataChannelMessage.t3 == '') {
```
Comment (sirknightj) [inline examples/viewer.js:527]: ===

File: examples/viewer.js (hunk 2 above)
Comment (sirknightj) [reply examples/viewer.js:1058]: Also may be able to simplify to just `if (metrics.master[key]) {`

File: examples/viewer.js [hunk 5]
```diff
@@ -646,3 +1022,76 @@ function calcStats(stats, clientId) {
...
+    let viewerOrder = ['signaling', 'describeChannel', 'describeMediaStorageConfiguration', 'channelEndpoint', 'iceServerConfig', 'connectAsViewer', 'setupMediaPlayer', 'waitTime',
+                    'offAnswerTime', 'iceGathering', 'peerConnection', 'dataChannel', 'ttffAfterPc', 'video'];
+    let masterOrder = ['signaling', 'describeChannel', 'channelEndpoint', 'iceServerConfig', 'getToken', 'createChannel', 'connectAsMaster', 'waitTime', 
+                    'offAnswerTime', 'iceGathering', 'peerConnection', 'dataChannel', 'ttffAfterPc'];
+    let container = document.getElementById('timeline-chart');
+    let rowHeight = 45;
+    let containerHeight = rowHeight;
+    let chart = new google.visualization.Timeline(container);
+    let dataTable = new google.visualization.DataTable();
+    let minTime = metrics.master.signaling.startTime < metrics.viewer.signaling.startTime ? metrics.master.signaling.startTime : metrics.viewer.signaling.startTime;
```
Comment (sirknightj) [inline examples/viewer.js:1047]: Can just use math.min()?

File: examples/viewer.js (hunk 4 above)
Comment (sirknightj) [reply examples/viewer.js:527]: let's include the PR explanation of this "The datachannel first sends a message of the following format with t1 attached, to which the master responds back with..." as a comment?

File: examples/viewer.js [hunk 6]
```diff
@@ -36,16 +38,239 @@ let droppedFramePerArray = [];
 let videoBitRateArray = [];
 let audioRateArray = [];
 let timeArray = [];
+let chartHeight = 0;
+
+let metrics = {
```
Comment (sirknightj) [inline examples/viewer.js:43]: Nice work!

Let's add a comment explaining that these metrics are received over datachannel msgs from the C SDK sample (from pr description).

Comment (niyatim23) [reply examples/viewer.js:43]: I have explained the flow but not mentioned the C SDK in particular. I can add that

File: examples/viewer.js (hunk 4 above)
Comment (niyatim23) [reply examples/viewer.js:527]: Included in the description already

Comment (sirknightj) [reply examples/viewer.js]: When it's cloned locally, we don't really tend to look at the PR description anymore. mostly comments

**Review by sirknightj — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -369,6 +369,15 @@ An enum with the following values:
...
 * `VIEWER`
 
+## Analysis
+
+<details>
+  <summary>Metrics</summary>
+
+The DQP Metrics can be enabled in the sample application by checking a box before starting the viewer. This helps get a detailed breakdown of time-to-first-frame and all the processes and API calls on master and the viewer both. This is intended to be used with the KVS WebRTC C SDK running as the master and the JS SDK as the viewer. The WebRTC C SDK needs to be built with `cmake .. -DENABLE_SENDING_METRICS_TO_VIEWER=ON`. The master sends peer, ice-agent, signaling and data-channel metrics to the viewer which are plotted ~ 20 seconds after the viewer is started. Since the timeline plot is intended to understand the time-to-first-frame, the sample web page needs to be refreshed and the master needs to be restarted if a new / updated plot is needed.  
+
+</details>
```
Comment (disa6302) [inline README.md:379]: I think we should detail the list of metrics here and clarify what the start and end points for the calculations are.

File: examples/viewer.js [hunk 8]
```diff
@@ -36,16 +38,239 @@ let droppedFramePerArray = [];
...
+            color: '#90CAF9',
+        }, 
+        peerConnection: {
+            name: 'pc-establishment-viewer',
+            startTime: '',
+            endTime: '',
+            tooltip: 'Time taken to establish the peer connection on the viewer',
+            color: '#2196F3',
+        },
+        ttffAfterPc: {
```
Comment (disa6302) [inline examples/viewer.js:122]: Since this is a metric we calculate in our samples on the C side, what would happen if customers do not include this? How would it show up in the graph?

Comment (niyatim23) [reply examples/viewer.js:122]: The customers cannot control excluding it, however, if the values are somehow altered in such a way that they cannot be displayed / are invalid, I have checks to skip them in the function at the end

File: README.md (hunk 7 above)
Comment (niyatim23) [reply README.md]: Added that now

File: examples/viewer.js (hunk 8 above)
Comment (disa6302) [reply examples/viewer.js:122]: I mean, ttff is something that is added in our C samples (it is not within the SDK core). Customers have often excluded the calculation for this when they write their own sample. These are still metrics coming from C SDK right?

File: examples/viewer.js [hunk 9]
```diff
@@ -36,16 +38,239 @@ let droppedFramePerArray = [];
...
+            endTime: '',
+            tooltip: 'Time to first frame after the master\'s peer connection has been established',
+            color: '#2196F3',
+        }
+    }
+};
+
+let dataChannelLatencyCalcMessage = {
+    content: 'Opened data channel by viewer',
+    timestamp1: '',
```
Comment (disa6302) [inline examples/viewer.js:241]: Lets be more descriptive in the naming here instead of timestamp1,2,3...

File: examples/viewer.js [hunk 10]
```diff
@@ -36,16 +38,239 @@ let droppedFramePerArray = [];
...
-async function startViewer(localView, remoteView, formValues, onStatsReport, onRemoteDataMessage) {
+async function startViewer(localView, remoteView, formValues, onStatsReport, remoteMessage) {
     try {
         console.log('[VIEWER] Client id is:', formValues.clientId);
 
         viewer.localView = localView;
         viewer.remoteView = remoteView;
 
+        viewer.remoteView.addEventListener('loadeddata', () => {
+            metrics.viewer.video.endTime = Date.now();
```
Comment (disa6302) [inline examples/viewer.js:256]: Is the intent here to record time to first decoded frame? If so, the naming seems misleading. Can we be specific with the naming here?

Also, this event should be listened to only if DQP is enabled or timeline display is enabled, alright if we dont want to have that check, but I dont think we need to do this needlessly either.

File: examples/viewer.js [hunk 11]
```diff
@@ -223,18 +473,127 @@ async function startViewer(localView, remoteView, formValues, onStatsReport, onR
...
+                if (formValues.enableDQPmetrics) {
+                    dataChannelLatencyCalcMessage.timestamp1 = Date.now();
+                    dataChannelObj.send(JSON.stringify(dataChannelLatencyCalcMessage));
+                } else {
+                    dataChannelObj.send("Opened data channel by viewer");
+                }
             };
             // Callback for the data channel created by viewer
-            dataChannelObj.onmessage = onRemoteDataMessage;
+            let onRemoteDataMessageViewer = (message) => {
```
Comment (disa6302) [inline examples/viewer.js:519]: What happens if the C SDK does not enable this? As I understand, master end sending is disabled by default in C SDK. I think we should have a checkbox specific to this. It should be separate from the DQP checkbox. C SDK customers could customize the data channel application to not send these or even not enable it, but still want viewer side metrics.

File: examples/viewer.js (hunk 9 above)
Comment (niyatim23) [reply examples/viewer.js]: Done!

File: examples/viewer.js (hunk 10 above)
Comment (niyatim23) [reply examples/viewer.js]: Changed the name. Introduced a new flag which is checked when this feature is to be used. Added a screenshot in the description

File: examples/viewer.js (hunk 11 above)
Comment (niyatim23) [reply examples/viewer.js:530]: Introduced a new checkbox and decoupled this from DQP. Also tested locally in all modes (DQP + Datachannel benchmarking, only DQP, only Datachannel benchmarking, none) and everything works fine
Screenshot 2023-12-19 at 9 06 27 PM

File: examples/index.html [hunk 12]
```diff
@@ -177,6 +178,18 @@ <h4>Amazon KVS WebRTC DQP</h4>
                 </div>
             </div>
 
+            <h4>Amazon KVS WebRTC Datachannel Benchmarking</h4>
+            <div class="form-group">
+                <div class="form-check form-check-inline">
+                    <input class="form-check-input" type="checkbox" id="enableDatachannelBenchmarking" value="enableDatachannelBenchmarking">
+                    <label for="enableDatachannelBenchmarking" class="form-check-label">Enable KVS WebRTC Datachannel Benchmarking <small>(Master + Viewer)</small></label>
```
Comment (disa6302) [inline examples/index.html:185]: The line suggests we are benchmarking data channel. Can we change to `Enable C SDK master profile timeline`? This is more accurate. Lets change the label too to something similar like `enableCMasterProfileTimeline`

**Review by disa6302 — APPROVED** [review]

File: examples/viewer.js (hunk 8 above)
Comment (niyatim23) [reply examples/viewer.js:122]: ttff is not coming from the C SDK, it is viewer specific. It is the first-frame - viewerrButtonPressed

File: examples/index.html (hunk 12 above)
Comment (niyatim23) [reply examples/index.html]: Renamed it to enableProfileTimeline as we are plotting master and viewer both
